pkg/block/blocktest/multipart.go

```diff
@@ -0,0 +1,315 @@
+
+	// verify no parts to list
+	verifyListInvalid(t, ctx, adapter, obj, uploadID)
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	if err != nil {
+		if adapter.BlockstoreType() == block.BlockstoreTypeGS {
+			require.ErrorContains(t, err, "part")
+		} else {
+			require.ErrorContains(t, err, "404")
+		}
+	}
+}
+
+// Test of the Multipart Copy API
+func testAdapterCopyPart(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	parts, full := createMultipartContents()
+	obj, objCopy := objPointers(storageNamespace)
+	uploadMultiPart(t, ctx, adapter, obj, parts)
+
+	// Begin Multipart Copy
+	resp, err := adapter.CreateMultiPartUpload(ctx, objCopy, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+	multiParts, err := copyPart(t, ctx, adapter, obj, objCopy, resp.UploadID)
+	if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+		require.ErrorContains(t, err, "not implemented")
+		return
+	}
+	_, err = adapter.CompleteMultiPartUpload(ctx, objCopy, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	require.NoError(t, err)
+
+	getAndCheckContents(t, ctx, adapter, full, objCopy)
+}
+
+// Test of the Multipart CopyRange API
+func testAdapterCopyPartRange(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	parts, full := createMultipartContents()
+	obj, objCopy := objPointers(storageNamespace)
+	uploadMultiPart(t, ctx, adapter, obj, parts)
+
+	// Begin Multipart Copy
+	resp, err := adapter.CreateMultiPartUpload(ctx, objCopy, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+	multiParts := copyPartRange(t, ctx, adapter, obj, objCopy, resp.UploadID)
+	if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+		// not implemented in Azure
+		require.Nil(t, multiParts)
+		return
+	}
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, objCopy, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	require.NoError(t, err)
+
+	getAndCheckContents(t, ctx, adapter, full, objCopy)
+}
+
+func createMultipartContents() ([][]byte, []byte) {
+	parts := make([][]byte, multipartNumberOfParts)
+	var partsConcat []byte
+	for i := 0; i < multipartNumberOfParts; i++ {
+		parts[i] = randstr.Bytes(multipartPartSize + i)
+		partsConcat = append(partsConcat, parts[i]...)
+	}
+	return parts, partsConcat
+}
+
+func uploadMultiPart(t *testing.T, ctx context.Context, adapter block.Adapter, obj block.ObjectPointer, parts [][]byte) {
+	t.Helper()
+	resp, err := adapter.CreateMultiPartUpload(ctx, obj, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+
+	multiParts := uploadParts(t, ctx, adapter, obj, resp.UploadID, parts)
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	require.NoError(t, err)
+}
+
+// List parts on non-existing part
```

Comment:
what is a non-existing part? 
If these tests list parts on a non-existing uploadID why is it tested in every test?